### ISSUE TITLE
sciml_train: unexpected: call back args out of sync with loss() return values

### ISSUE THREAD

**Issue by cems2** [JuliaDiffEq/DiffEqFlux.jl#166]
an unexpected result : the args sent to callback (params,loss) are not self-consistent.  (Since there is a cost associated with making them consistent the solution here is not to pay that cost but rather to document this inconsistency.)

**Expected behavior:**
 if you set loss to return  the scalar loss then   two args will also get sent to the callback. (params, loss).  You would expect that if you use those params to compute the loss it would match the loss being sent to cb.

**Actual Behavior**
 The callback gets a loss and a set of parameters as expected.  **But the parameters sent to cb() are not the ones used to compute the associated loss value  sent to cb()**.  The parameters correspond to the  _next iteration_ not the current iteration used to compute the loss value.  Surprise!

**Code to reproduce:**
This is not complete, just the snippet to explain the output (see end for full code)
```
prob = ODEProblem(f,u0,(0.0f0,1.0f0),p)

function loss(p)
   pred = concrete_solve(prob,Tsit5(),u0,p, saveat=0.01)
   sum(abs, (pred.-u0))
end

function cb1(args...)
   println("args:",args)
   false
end

res = DiffEqFlux.sciml_train(loss,p,ADAM(0.05),maxiters=30,cb=cb1)
```
then when I run it I get the following output
```
<snipped to just the last two outputs>
args:(Float32[0.7163402, 0.6396542, -0.548202, -0.09538113, -0.27492765, -0.40611613], 6.6804714f0)
args:(Float32[0.7258747, 0.64094096, -0.5339987, -0.090279065, -0.26264328, -0.4026812], 3.8790169f0)

```
now lets manually compute the loss for the Second-to-last set of parameters printed out by cb()
```
julia> loss(Float32[0.7163402, 0.6396542, -0.548202, -0.09538113, -0.27492765, -0.40611613])
3.879016f0

```

Say what? you would expect second-to-last parameters would correspond to the second to last loss 6.6804714f0 sent to cb()  but instead it's the loss from the next iteration of cb().  3.879016f0

**thus the pair (parameters, loss) sent to cb are not consistent with each other.**

**Amusing Pitfall**
also I want to offer a word to of caution about an amusing pitfall one can very quickly encounter when developing in Juno with regard to the callback function.  This is absolutely expected given how julia works.  It's just super easy to trigger.

I first wrote:
```
function cb1(P,L)
   println("loss: ",L, " params:", P)
   false
end
```
that works just fine but then I deleted it and changed this to
```
function cb1(args...)
   println("args:",args)
   false
end
```

And surprise! if you rerun the sciml_train( ...  cb=cb1) it calls the orginal form of cb1  not the revised one.

But then if I change the loss function to return an additional item it's return tuple, it calls the new form of cb1.

Again this is correct behavior for julia (not python) but something that doesn't happen often can easily happen when you are tinkering with the number of things loss returns and using Juno to re-run things without stopping the REPL.

**fully runable code**
```
using Flux #for ADAM #DiffEqFlux,
using Optim # for BFGS
#using DifferentialEquations
using OrdinaryDiffEq
using DiffEqFlux

model1 = FastChain(FastDense(2,2))
model2 = FastChain(FastDense(2,2))
nontrainable = initial_params(model2)
p = initial_params(model1)

f(x,p,t) = model2(model1(x,p),nontrainable)

u0 = Float32[1.0,1.0]

prob = ODEProblem(f,u0,(0.0f0,1.0f0),p)

function loss(p)
   pred = concrete_solve(prob,Tsit5(),u0,p, saveat=0.01)
   sum(abs, (pred.-u0))
end

function cb1(args...)
   println("args:",args)
   false
end

res = DiffEqFlux.sciml_train(loss,p,ADAM(0.05),maxiters=30,cb=cb1)
```

**Comment by lungd** [comment]
Swapping the invocations of Flux.Optimize.update!() and cb() inside sciml_train() should fix this issue.

The following code with a custom sciml_train function works as you would expect

```
using Flux #for ADAM #DiffEqFlux,
using Optim # for BFGS
#using DifferentialEquations
using OrdinaryDiffEq
using DiffEqFlux
using Juno


struct NullData end
const DEFAULT_DATA = Iterators.cycle((NullData(),))
Base.iterate(::NullData, i=1) = nothing
Base.length(::NullData) = 0

get_maxiters(data) = Iterators.IteratorSize(typeof(DEFAULT_DATA)) isa Iterators.IsInfinite ||
                     Iterators.IteratorSize(typeof(DEFAULT_DATA)) isa Iterators.SizeUnknown ?
                     typemax(Int) : length(data)


function my_sciml_train(loss, _θ, opt, _data = DEFAULT_DATA;
                     cb = (args...) -> (false),
                     maxiters = get_maxiters(data))

  # Flux is silly and doesn't have an abstract type on its optimizers, so assume
  # this is a Flux optimizer
  θ = copy(_θ)
  ps = Flux.params(θ)

  if _data == DEFAULT_DATA && maxiters == typemax(Int)
    error("For Flux optimizers, either a data iterator must be provided or the `maxiters` keyword argument must be set.")
  elseif _data == DEFAULT_DATA && maxiters != typemax(Int)
    data = Iterators.repeated((), maxiters)
  elseif maxiters != typemax(Int)
    data = take(_data, maxiters)
  else
    data = _data
  end

  t0 = time()
  local x
  @progress for d in data
    gs = Flux.Zygote.gradient(ps) do
      x = loss(θ,d...)
      first(x)
    end
    cb_call = cb(θ,x...)
    if !(typeof(cb_call) <: Bool)
      error("The callback should return a boolean `halt` for whether to stop the optimization process. Please see the sciml_train documentation for information.")
    elseif cb_call
      break
    end


    Flux.Optimise.update!(opt, ps, gs)    ################   this line is now below cb()


  end

  _time = time()
  Optim.MultivariateOptimizationResults(opt,
                                        _θ,# initial_x,
                                        θ, #pick_best_x(f_incr_pick, state),
                                        first(x), # pick_best_f(f_incr_pick, state, d),
                                        maxiters, #iteration,
                                        maxiters >= maxiters, #iteration == options.iterations,
                                        false, # x_converged,
                                        0.0,#T(options.x_tol),
                                        0.0,#T(options.x_tol),
                                        NaN,# x_abschange(state),
                                        NaN,# x_abschange(state),
                                        false,# f_converged,
                                        0.0,#T(options.f_tol),
                                        0.0,#T(options.f_tol),
                                        NaN,#f_abschange(d, state),
                                        NaN,#f_abschange(d, state),
                                        false,#g_converged,
                                        0.0,#T(options.g_tol),
                                        NaN,#g_residual(d),
                                        false, #f_increased,
                                        nothing,
                                        maxiters,
                                        maxiters,
                                        0,
                                        true,
                                        NaN,
                                        _time-t0,)
end


model1 = FastChain(FastDense(2,2))
model2 = FastChain(FastDense(2,2))
nontrainable = initial_params(model2)
p = initial_params(model1)

f(x,p,t) = model2(model1(x,p),nontrainable)

u0 = Float32[1.0,1.0]

prob = ODEProblem(f,u0,(0.0f0,1.0f0),p)

function loss(p)
   pred = concrete_solve(prob,Tsit5(),u0,p, saveat=0.01)
   sum(abs, (pred.-u0))
end

function cb1(args...)
   println("args:",args)
   false
end

res = my_sciml_train(loss,p,ADAM(0.05),maxiters=30,cb=cb1)
```

**Comment by cems2** [comment]
comment.  Depending on Why CB returns false or True one might want to put the test for that return value before or after update.  Example:  typically the rationale would be that CB decides "hey this thing is minimized good enough and isn't making progress so let's stop now".  In that case the CB call goes before update and the test of the CB return value goes after update.

BUT... if CB's reason for stopping is "Holy crap! the loss function (or some regularizer) just shot up to the moon, I don't want to take this step!" then you'd want the check before update.

This latter use case I think will never be the case and could be handled other ways (having CB cache the last best values for example).  The former use case is thus the one I think should be the default.

Finally the docs may wish to point out that the loss and params reported in the last call to CB may be different that the parameter values (and loss) at the exit of Flux.Train! due to one last update.  This isn't undesirable.  It's just a mild surprise and much better than the original suprise of this issue.

**Comment by ChrisRackauckas** [comment]
With https://github.com/JuliaDiffEq/DiffEqFlux.jl/pull/172 it'll only update if it's continuing.

**Comment by cems2** [comment]
Right.  But no reason to throw out the last gradient?  it's info so the update probably will be one step better.

**Comment by ChrisRackauckas** [comment]
not necessarily if using a stochastic optimizer.

**Comment by cems2** [comment]
right.  that was my case #2.  Both are sensible. in retrospect I think the deteministic method is more satisfying, so go with it
